### PR TITLE
refactor code to MVP pattern

### PR DESCRIPTION
**Pull request by vuduychuong**

### CODE REVIEW

File: app/src/main/java/com/framgia/fel1/network/API.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+/**
+ * Created by vuduychuong1994 on 3/22/17.
+ */
+
+public abstract class API {
+    private static final String BASE_URL = "https://manh-nt.herokuapp.com/";
+    private static OkHttpClient.Builder sHttpClient = new OkHttpClient.Builder();
+
+    private static Retrofit.Builder builder =
```
Comment (tuanbaros) [inline app/src/main/java/com/framgia/fel1/network/API.java]: convention

File: app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+    private RecyclerView mRecyclerViewCategory;
+    private CategoryAdapter mCategoryAdapter;
+    private ArrayList<Category> mListCategory = new ArrayList<>();
+    private String mAuthToken;
+    private User mUser;
+    private MySqliteHelper mMySqliteHelper;
+    private Toast mToast;
+    private SharedPreferences mSharedPreferences;
+    private static boolean sIsCategoryLoad = false;
+    private boolean isLoadImage = false;
```
Comment (tuanbaros) [inline app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java]: convention

File: app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+        mTextViewEmail = (TextView) view.findViewById(R.id.text_show_user_email);
+        mRecyclerViewCategory = (RecyclerView) findViewById(R.id.listview_lesson_learned);
+        mRecyclerViewCategory.setLayoutManager(new LinearLayoutManager(getApplicationContext()));
+        mCategoryAdapter = new CategoryAdapter(this, mListCategory);
+        mRecyclerViewCategory.setAdapter(mCategoryAdapter);
+        mSharedPreferences = getSharedPreferences(Const.MY_PREFERENCE, Context.MODE_PRIVATE);
+        int id = mSharedPreferences.getInt(Const.ID, -1);
+        if (id != -1) mUser = mMySqliteHelper.getUser(id);
+        else finish();
+        if (!isLoadImage) if (InternetUtils.isInternetConnected(HomeActivity.this, false)) {
```
Comment (tuanbaros) [inline app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java]: refactor

File: app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+        if (!sIsCategoryLoad) {
+            showDialog();
+            mPresenter.getListCategory(mAuthToken);
+        }
+        mImageViewAvatar.setOnClickListener(this);
+    }
+
+    @Override
+    protected void onResume() {
+        Bitmap bitmap = BitmapUtil.decodeSampledBitmapFromFile(mUser.getAvatar(), 100, 100);
```
Comment (tuanbaros) [inline app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java]: hardcode

File: app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+        if (bitmap != null) mImageViewAvatar.setImageBitmap(bitmap);
+        super.onResume();
+    }
+
+    @Override
+    public void onClick(View v) {
+        switch (v.getId()) {
```
Comment (tuanbaros) [inline app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java]: use if

File: app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+            case R.id.image_show_user_avatar:
+                Intent intent = new Intent(HomeActivity.this, UpdateProfileActivity.class);
```
Comment (tuanbaros) [inline app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java]: suggest: startActivity(new Intent...)

File: app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+             .setPositiveButton(R.string.thoat,
+                     new DialogInterface.OnClickListener() {
+                         @Override
+                         public void onClick(
+                                 DialogInterface dialog, int which) {
+                             onSignOut();
+                         }
+                     })
+             .setNegativeButton(R.string.huy,
+                new DialogInterface.OnClickListener() {
```
Comment (tuanbaros) [inline app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java]: .setNegativeButton(R.string.huy, null);

File: app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+            super.onBackPressed();
+            sIsCategoryLoad = false;
+        } else {
+            mToast.show();
+        }
+    }
+
+    @Override
+    public void onPreExecute() {
+        if (sGetTag.equals(SIGOUT_TAG)) {
```
Comment (tuanbaros) [inline app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java]: 1 line

File: app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+        startActivity(intentLessonLearned);
+        mSharedPreferences = getSharedPreferences(Const.MY_PREFERENCE, Context.MODE_PRIVATE);
+        SharedPreferences.Editor editor = mSharedPreferences.edit();
+        editor.putInt(Const.CATEGORY_ID, mListCategory.get(position).getId());
+        editor.apply();
+    }
+
+    @Override
+    public void onBackPressed() {
+        if (mToast == null) mToast =
```
Comment (tuanbaros) [inline app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java]: refactor

File: app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+                e.printStackTrace();
+            }
+            return response;
+        }
+        return response;
+    }
+
+    @Override
+    public void onProgressUpdate(String response) {
+
```
Comment (tuanbaros) [inline app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java]: todo

File: app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+    }
+
+    @Override
+    public void onCancelled() {
+
```
Comment (tuanbaros) [inline app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java]: todo

File: app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java [hunk 12]
```diff
@@ -0,0 +1,372 @@
...
+    }
+
+    @Override
+    public void onPostExecute(String response) {
+        if (sGetTag == SIGOUT_TAG) {
```
Comment (tuanbaros) [inline app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java]: refactor

File: app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+                Intent intentWordList = new Intent(HomeActivity.this, WordListActivity.class);
+                startActivity(intentWordList);
+                break;
+            case R.id.nav_activities:
+                Intent intentActivities = new Intent(HomeActivity.this, UserActionActivity.class);
+                startActivity(intentActivities);
+                break;
+            case R.id.nav_signout:
+                showSignOutDialog();
+                break;
```
Comment (tuanbaros) [inline app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java]: default

File: app/src/main/java/com/framgia/fel1/presentation/home/HomePresenter.java [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+    public void getListCategory(String authToken) {
+        mCategoryRepository.getCategories(authToken, new CategoryDataSource.LoadCategoryCallback() {
+            @Override
+            public void onCategoryLoaded(List<Category> categoryList) {
+                mHomeView.showListCategory(categoryList);
+            }
+
+            @Override
+            public void onDataNotAvailable() {
+                Log.d(TAG, "data not avaiable");
```
Comment (tuanbaros) [inline app/src/main/java/com/framgia/fel1/presentation/home/HomePresenter.java]: remove log and TAG

File: app/src/main/java/com/framgia/fel1/data/source/category/remote/CategoryRemoteDataSource.java [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    @Override
+    public void getCategories(
+            final String authToken, @NonNull final LoadCategoryCallback callback) {
+        API.createService(APIService.class)
+           .getCategories(authToken)
+           .enqueue(new Callback<CategoryResponse>() {
+               @Override
+               public void onResponse(
+                       Call<CategoryResponse> call, Response<CategoryResponse> response) {
+                   callback.onCategoryLoaded(response.body().getCategoryList());
```
Comment (tuanbaros) [inline app/src/main/java/com/framgia/fel1/data/source/category/remote/CategoryRemoteDataSource.java]: check respone null

File: app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java [hunk 16]
```diff
@@ -0,0 +1,359 @@
...
+        mRecyclerViewCategory = (RecyclerView) findViewById(R.id.listview_lesson_learned);
+        mRecyclerViewCategory.setLayoutManager(new LinearLayoutManager(getApplicationContext()));
+        mCategoryAdapter = new CategoryAdapter(this, mListCategory);
+        mRecyclerViewCategory.setAdapter(mCategoryAdapter);
+        mSharedPreferences = getSharedPreferences(Const.MY_PREFERENCE, Context.MODE_PRIVATE);
+        int id = mSharedPreferences.getInt(Const.ID, -1);
+        if (id != -1) mUser = mMySqliteHelper.getUser(id);
+        else finish();
+        if (!mIsLoadImage && InternetUtils.isInternetConnected(HomeActivity.this, false)) {
+            mIsLoadImage = !mIsLoadImage;
```
Comment (tuanbaros) [inline app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java]: mIsLoadImage  = true;

File: app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java [hunk 17]
```diff
@@ -0,0 +1,359 @@
...
+            if (mUser.getAvatar() != null || !mUser.getAvatar().equals("")) {
```
Comment (tuanbaros) [inline app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java]: use TextUtil

File: app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java [hunk 18]
```diff
@@ -0,0 +1,359 @@
...
+        if (sGetTag.equals(SIGOUT_TAG)) {
+            String url = NetwordConst.URL_API_SIGNOUT + "?" +
+                    Const.AUTH_TOKEN + "=" +
+                    mUser.getAuthToken();
+            try {
+                response = HttpRequest.postJsonRequest(url, null, NetwordConst.METHOD_DELETE);
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+            return response;
```
Comment (tuanbaros) [inline app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java]: remove

File: app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java [hunk 19]
```diff
@@ -0,0 +1,352 @@
...
+    private void showSignOutDialog() {
+        new AlertDialog.Builder(this).setTitle(R.string.infor).setMessage(R.string.confirrn_signout)
+             .setPositiveButton(R.string.thoat,
+                     new DialogInterface.OnClickListener() {
+                         @Override
+                         public void onClick(
+                                 DialogInterface dialog, int which) {
+                             onSignOut();
+                         }
+                     }).setNegativeButton(R.string.huy, null).show();
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/fel1/presentation/home/HomeActivity.java]: sửa string

**Pull request merged**